File: example/java/frugal-lib/com/workiva/frugal/Transport.java
package com.workiva.frugal;

import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;
import org.apache.thrift.transport.TTransportFactory;

/**
 * Transport wraps a Thrift TTransport which supports pub/sub.
 */
public interface Transport {

    /**
     * Opens the Transport to receive messages on the subscription.
     *
     * @param topic the pub/sub topic to subscribe to.
     * @throws TTransportException
     */
    void subscribe(String topic) throws TTransportException;

    /**
     * Closes the Transport to stop receiving messages on the subscription.
     *
     * @throws TTransportException
     */
    void unsubscribe() throws TTransportException;

    /**
     * Prepares the Transport for publishing to the given topic.
     *
     * @param topic the pub/sub topic to publish on.
     */
    void preparePublish(String topic);

    /**
     * Returns the wrapped Thrift TTransport.
     *
     * @return wrapped transport.
     */
    TTransport thriftTransport();

    /**
     * Wraps the underlying TTransport with the TTransport returned by the given
     * TTransportFactory.
     *
     * @param proxy the factory to wrap the transport with.
     */
    void applyProxy(TTransportFactory proxy);

}

File: Godeps/_workspace/src/github.com/Workiva/frugal-go/transport.go
package frugal

import "git.apache.org/thrift.git/lib/go/thrift"

// FTransportFactory is responsible for creating new Frugal FTransports.
type FTransportFactory interface {
	// GetTransport creates a new FTransport.
	GetTransport() FTransport
}

// FTransport wraps a Thrift TTransport which supports pub/sub.
type FTransport interface {
	// Subscribe opens the FTransport to receive messages on the subscription.
	Subscribe(string) error

	// Unsubscribe closes the FTransport to stop receiving messages on the
	// subscription.
	Unsubscribe() error

	// PreparePublish prepares the FTransport for publishing to the given
	// topic.
	PreparePublish(string)

	// ThriftTransport returns the wrapped Thrift TTransport.
	ThriftTransport() thrift.TTransport

	// ApplyProxy wraps the underlying TTransport with the TTransport returned
	// by the given TTransportFactory.
	ApplyProxy(thrift.TTransportFactory)
}
